### PR TITLE
Allow closures to take both struct- and class-based capture environments

### LINKED ISSUE #16895: Compiler crash when compiling nested lambda inside local function capturing both a local in the local function and a parameter >= 2 scopes outside local function
Here is the simplest repro I can find:
```csharp
        [Fact]
        public void CaptureVarNestedLambdaSkipScope()
        {
            var src = @"
using System;
class C
{

    public static void Main()
    {
        var d = """";
        {
            int x = 0;
            void M()
            {
                if (d != null)
                {
                    Action a = () => x++;
                    a();
                }
            }
            M();
        }
    }
}";
            CompileAndVerify(src, options: TestOptions.DebugExe, expectedOutput: "1");
        }
```

### PR DESCRIPTION
**Pull request by agocke**
Lambda rewriting currently allows frames to be structs in the instance
where

  1. The variables in the frame are not captured by any closures
  converted to delegates.

  2. A closure nested inside the struct closure captures a separate
  variable that does require a class frame.

This poses a problem when the "outer" closure is in the same frame as
the variable that requires a class closure. The problem is that we treat
scopes and frames as essentially the same thing -- if any scope is
perceived to require a particular environment type, all captured
variables in that scope will be captured on that frame and all closures
will be lowered onto that frame.

This creates a conflict between class-based environment capturing, where
we want to capture the frame pointer as a field, and struct-based frame
capturing, where we want to add arguments as ref parameters. Doing both
results in breaking the delegate contract (no signature rewriting) or
losing required struct arguments to intermediate closures.

Theoretically we could support this scenario -- "closure skipping" in
this manner could be supported by analyzing not just what frame is
captured by a particular closure, but also what inside the frame is
captured and what in any parent frames are captured. However, at the
moment we have essentially one bit of information for each scope --
just whether or not it captures its containing scope. Thus, we can't
differentiate what's required by which closure and so we can't limit
passing the frames only to the closures that require them.

The simple fix here is to require that every frame that it is captured
by a frame that must be a class is itself required to be a class
(transitively).

Fixes #16895

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 1]
```diff
@@ -1453,6 +1431,31 @@ private DebugId GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, int clos
             return synthesizedMethod;
         }
 
+        /// <summary>
+        /// Builds a list of all the struct-based closure environments that will
+        /// need to be passed as arguments to the method.
+        /// </summary>
+        private ImmutableArray<TypeSymbol> GetStructClosures(LambdaFrame containerAsFrame, BoundNode lambdaScope)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Note: this refactoring should be equivalent, I just had trouble understanding the previous logic.

**Comment by agocke** [comment]
@dotnet/roslyn-compiler For review

**Comment by agocke** [comment]
@dotnet/roslyn-compiler Ping for review

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs [hunk 2]
```diff
@@ -342,6 +342,23 @@ internal void ComputeLambdaScopesAndFrameCaptures()
...
+                foreach (var scope in ScopesThatCantBeStructs)
+                {
+                    var curScope = scope;
+                    while (NeedsParentFrame.Contains(curScope))
+                    {
+                        if (!ScopeParent.TryGetValue(curScope, out curScope))
+                        {
+                            break;
+                        }
+                        moreNoStructScopes.Add(curScope);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs]: >moreNoStructScopes.Add(curScope); [](start = 24, length = 33)

It feels like we can jump out of this loop as soon as curScope is already in ScopesThatCantBeStructs or in moreNoStructScopes. It might make sense to change moreNoStructScopes to be a HashSet.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 3]
```diff
@@ -1324,35 +1324,13 @@ private DebugId GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, int clos
...
-                        }
-                        if (found)
-                        {
-                            continue;
-                        }
-                    }
-                    // can happen when scope no longer needs parent frame, or we're at the outermost level and the "parent frame" is top level "this".
-                    lambdaScope = null;
+                    // Lower directly onto the containing type
                     containerAsFrame = null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: >containerAsFrame = null; [](start = 20, length = 24)

It looks like this statement is no longer needed.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: This doesn't look equivalent, the lambdaScope should probably be ```out``` for that.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: The same is for containerAsFrame parameter. Should probably be ```ref```.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Unless we want a behavior change here, I suggest to undo this manual refactoring and go with IDE refactoring as a separate PR.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs [hunk 4]
```diff
@@ -105,6 +105,159 @@ void Local(string s = nameof(Local))
...
+                int x = 0;
+                bool Local(int y) => x == 0 && args == null && y == 0;
+                Local(0);
+            }
+        }
+    }
+}";
+            CompileAndVerify(src);
+
+        }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs]: >} [](start = 8, length = 1)

There is an extra empty line added after this function.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 5]
```diff
@@ -1453,6 +1433,45 @@ private DebugId GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, int clos
...
+                    structClosureParamBuilder.Add(containerAsFrame);
+                }
+                if (_analysis.NeedsParentFrame.Contains(lambdaScope))
+                {
+                    bool found = false;
+                    while (_analysis.ScopeParent.TryGetValue(lambdaScope, out lambdaScope))
+                    {
+                        if (_frames.TryGetValue(lambdaScope, out containerAsFrame))
+                        {
+                            found = true;
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Nit: This seems to be one of those cases where "goto" would make it easier to read. You need "found" just to implement 2-level `continue`.

**Comment by agocke** [comment]
To elaborate on the problem here, let's consider the example:

```csharp
using System;
class C
{
    public void M()
    {
        int x = 0;
        {
            int y= 0;
            void Local()
            {
                if (d == 0)
                {
                    Action a = () => y++;
                    a();
                }
            }
            Local();
        }
    }
}
```

The current problem in the compiler is in the code now in `GetStructClosures`. The previous implementation only built struct closure if the closure was being lowered onto a struct frame.

However, in the example, `x` is being captured by a lambda, meaning that `x` must be in a class frame. Since `M` lives in the same scope it also lives in the same frame and contains the lambda capturing `x`, meaning that it must live in a class frame as well. Of course, it is not converted to a delegate, nor does it capture any variables that are captured by closures converted to delegates, so its signature is free to be rewritten to take variables which are not captured by closures converted to delegates as struct environment parameters.

Here is the expected lowering for the previous example:

```csharp
void M()
{
    var env1 = new Env1();
    env1.x = 0;
    var env2 = new Env2();
    env2.y = 0;
    env2.<>_L(ref env1);
}

struct Env1
{
    public int x;
}

class Env2
{
    public int y;
    public void <>_L(ref Env1 env1)
    {
        if (env1.x == 0)
        {
            var env3 = new Env3();
            env3.env2 = this;
            Action a = env3.<>_anon;
            a();
        }
    }
}

class Env3
{
     Env2 env2;
     public void <>_anon() => this.env2.y++;
}
```

The problem comes when calculating the struct frames needed to be passed to `Local`. In the current implementation, unless the "container" is a struct, no struct frames are added. If we fix that to check for struct frames that are required as parent frames, regardless of the type of container we run into another problem.

In this case, the lambda also includes struct frames in its parent "context", but adding *any* struct frames to a closure that is converted to a delegate is incorrect.

The previous problem can be solved by a check to see if the closure is converted to a delegate. This works because of something we already know from the analysis: if we assume that struct frames are only created when they contain variables not captured by closures converted to delegates, we can assume that any struct frames in scope do not contain variables that are used inside closures converted to delegates, even if those frames are in the "parent context" of the closure.

In my opinion, a better system for keeping track of this information is to list exactly the captured variables and closures required for each closure. If the set of variables captured by a nested closure is disjoint from one of its higher-level parent closures, that closure never needs to be captured -- and this can be verified directly.

**Comment by VSadov** [comment]
The sad part is that the outer frame does not even need to be captured by reference. It is semantically  readonly and as such does not need to be captured transitevely.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs (hunk 3 above)
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: `containerAsFrame` is used later in the method, I think it's still necessary.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs (hunk 1 above)
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Since my change, this refactoring is no longer intended to be equivalent. There is change here.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 6]
```diff
@@ -1324,35 +1324,15 @@ private DebugId GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, int clos
             if (_analysis.LambdaScopes.TryGetValue(node.Symbol, out lambdaScope))
             {
                 containerAsFrame = _frames[lambdaScope];
-                var structClosureParamBuilder = ArrayBuilder<TypeSymbol>.GetInstance();
-                while (containerAsFrame != null && containerAsFrame.IsValueType)
+                structClosures = _analysis.MethodsConvertedToDelegates.Contains(node.Symbol)
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: So, the actual fix seems to be just this condition?

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Would there be a problem with async/iterator methods?

---
In reply to: [100203934](https://github.com/dotnet/roslyn/pull/16908#discussion_r100203934) [](ancestors = 100203934)

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs [hunk 7]
```diff
@@ -73,7 +73,15 @@ internal sealed class Analysis : BoundTreeWalkerWithStackGuardWithoutRecursionOn
             public readonly HashSet<MethodSymbol> MethodsConvertedToDelegates = new HashSet<MethodSymbol>();
 
             /// <summary>
-            /// Any scope that a method in <see cref="MethodsConvertedToDelegates"/> closes over. If a scope is in this set, don't use a struct closure.
+            /// True if the method signature can't be rewritten to contain ref/out parameters.
+            /// </summary>
+            public bool CantTakeRefParameters(MethodSymbol closure) => closure.IsAsync
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs]: >CantTakeRefParameters [](start = 24, length = 21)

Minor: It is probably better to remove negation form the name ("CanTakeRefParameters") and adjust the returned value accordingly.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 8]
```diff
@@ -1324,35 +1324,15 @@ private DebugId GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, int clos
             if (_analysis.LambdaScopes.TryGetValue(node.Symbol, out lambdaScope))
             {
                 containerAsFrame = _frames[lambdaScope];
-                var structClosureParamBuilder = ArrayBuilder<TypeSymbol>.GetInstance();
-                while (containerAsFrame != null && containerAsFrame.IsValueType)
+                structClosures = _analysis.CantTakeRefParameters(node.Symbol)
+                    ? ImmutableArray<TypeSymbol>.Empty
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: >ImmutableArray<TypeSymbol>.Empty [](start = 22, length = 32)

It looks like other branches assign ```default(ImmutableArray<TypeSymbol>)```. Is ```ImmutableArray<TypeSymbol>.Empty``` used to communicate different information? Consider standardizing on the same value otherwise.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 9]
```diff
@@ -1453,6 +1433,45 @@ private DebugId GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, int clos
...
+        {
+            var structClosureParamBuilder = ArrayBuilder<TypeSymbol>.GetInstance();
+            while (containerAsFrame?.IsValueType == true ||
+                   _analysis.NeedsParentFrame.Contains(lambdaScope))
+            {
+                if (containerAsFrame?.IsValueType == true)
+                {
+                    structClosureParamBuilder.Add(containerAsFrame);
+                }
+                if (_analysis.NeedsParentFrame.Contains(lambdaScope) && FindParentFrame())
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: >FindParentFrame() [](start = 72, length = 17)

It feels that the code would be more readable (easier to follow) if lambdaScope and containerAsFrame were explicitly passed by reference.

**Comment by agocke** [comment]
test windows_eta_open_prtest please

**Comment by agocke** [comment]
@dotnet/roslyn-compiler Ping for re-review.
